Support for gzipped fastq formats

mvdbeek opened the pull request:
This is a collection of work from @ashvark, @pvanheus and perhaps others that I'm not aware of, and I think this has been implemented one way or the other by different galaxy admins (I think @hrhotz mentioned this, and various admins in france) and I think it's time we get one implementation into galaxy so that we can implement and rely on compressed fastq handling in tools.

The approach is that for each fastq variant we have the corresponding gzip compressed format, and for each compressed variant we have one converter (we may be able to compact this into one).

With the current implementation, users can upload a gzip-compressed fastq file, which will be properly sniffed as fastq.gz and will not be uncompressed. Tools that can deal with compressed fastq files can allow fastq.gz as input format, in which case the compressed fastq file can be used without decompression.
If the tool only allows uncompressed fastq, the file will go through the dataset converter.
~~By selecting on of the uncompressed fastq formats at upload, the files will still be unpacked.~~

I think this is not a big departure from the current user experience, and should help scale Galaxy in terms of storage use. I'm happy to make changes, if we can agree on one direction.

usablegarlic committed:
Changes for Fastq Compression.

usablegarlic committed:
Updated sequence.py and datatypes_conf.xml

ashvark committed:
Changes to keep the compressed fastq file as it is with out decompressing.

mvdbeek committed:
For fastq.gz support inherit from Fastq and Binary

and enhance FastQ format to be gzip aware.

pvanheus commented:
This is, in my mind, a better approach than have a "compressed" metadata flag. @zipho has done work at SANBI on supporting native compressed fastq. This can probably be contributed towards this proposed design.

bgruening commented:
@mvdbeek a tool test that triggers the automatic conversion would be neat.

zipho commented:
@mvdbeek We have for the most part, locally at SANBI (UWC) implemented similar functionality by creating a (fastq) Gz datatype and for our purposes, made also some changes in the upload.py (tools/data_source) to handle uploading of compressed fastq files. 

One issue though has been that some tools evaluate the extension of a file (.gz) to determine if that file is compressed or not and therefore storing datasets internally as .dat proved to present some challenges for certain tools.

mvdbeek commented:
Hi @zipho 

> made also some changes in the upload.py (tools/data_source) to handle uploading of compressed fastq files.

That's an implementation detail I have solved by inheriting from Binary and registering fastq.gz as a compressed datatype, which is enough for fastq.gz to not be uncompressed.

> One issue though has been that some tools evaluate the extension of a file (.gz) to determine if that file is compressed or not and therefore storing datasets internally as .dat proved to present some challenges for certain tools.

There is quite a number of tools that require a specific extension, but I think current iuc practice is to use symlinks in the tools `command` attribute to make these tools happy. Having a separate fastq.gz allows to ask whether a file is compressed and to adjust the command section accordingly.

In a future step I think galaxy could default to linking in files based on their datatypes extension, so that this becomes unnecessary, but that would be a bigger change, I think.
I have also come across tools that even need a specific filename (sigh), so I think some symlinking from the tool author side may be necessary even if we are going that direction.

zipho commented:
@mvdbeek I agree.

> There is quite a number of tools that require a specific extension, but I think current iuc practice is to use symlinks in the tools command attribute to make these tools happy. Having a separate fastq.gz allows to ask whether a file is compressed and to adjust the command section accordingly.

We ended up following the same practise as iuc on many of our tools. The winner certainly would be for galaxy to default to linking files based on extension.

mvdbeek commented:
> @mvdbeek a tool test that triggers the automatic conversion would be neat.

Thanks for asking for this, I discovered that while fastqsanger.gz is automatically converted to fastqsanger, fastq.gz is also valid fastq and is therefor not converted. I guess I need to create a BaseFastq class from which Fastq and FastqGz can inherit.

mvdbeek committed:
Create a BaseFastq class and add test tools

Having a BaseFastq class allows Fastq and FastqGz classes to inherit from
BaseFastq. FastqGz is therfor not a Fastq datatype and goes through the
converter.

Also add 2 tools to demonstrate that compressed fastq will be converted to
uncompressed fastq if the tool specifies `format="fastq"`, while
`format="fastq.gz"` leaves files compressed.

The tool test can be run with:
```
planemo test --galaxy_root .  test/functional/tools/compressed_fastq_no_conversion.xml
planemo test --galaxy_root .  test/functional/tools/compressed_fastq_conversion.xml
```

Note that fastq conversion appears to required a galaxy user session, and
planemo testing fails. Interactive testing with `planemo serve` is not affected
and works fine.

bgruening commented on lib/galaxy/datatypes/converters/fastqcssangergz_to_fastqcssanger.xml:
```diff
@@ -0,0 +1,13 @@
+<tool id="CONVERTER_fastqcssangergz_to_fastqcssanger" name="Convert fastqcssanger.gz files to fastqcssanger" version="1.0.0" hidden="true">
+  <command>gzip -dcf $input1 > $output1</command>
```
Should we add a requirement for all these tools?

mvdbeek replied:
Hmm, even the busybox and ubuntu docker base images come with gzip, os x has a system gzip ... I doubt we could satisfy a dependency or run galaxy without gzip.
Alternatively I could write a python script to do the decompression, if that's a concern.

bgruening replied:
Ok fair enough, I was more coming from the best-practice side, where I think all tools should be annotated even the python tools.

mvdbeek replied:
I understand, and normally I agree, but it's not so easy to get a requirement for gzip.

bgruening commented on lib/galaxy/datatypes/converters/fastqgz_to_fastq.xml:
```diff
@@ -0,0 +1,13 @@
+<tool id="CONVERTER_fastqgz_to_fastq" name="Convert fastq.gz files to fastq" version="1.0.0" hidden="true">
+  <command>gzip -dcf $input1 > $output1</command>
```
quoting all paths

bgruening commented on lib/galaxy/datatypes/sniff.py:
```diff
@@ -198,7 +198,12 @@ def get_headers( fname, sep, count=60, is_multi_byte=False ):
+    if compressed:
+        in_file = gzip.GzipFile(fname, 'r')
+    else:
+        in_file = open(fname, 'rt')
```
Do we need to close this file handle?

mvdbeek replied:
quickly scanning through the datatypes, some functions do and some don't. I'll add this.

bgruening reviewed (commented):
@mvdbeek I like this very much! Thanks!

mvdbeek committed:
Quote paths in converters

mvdbeek committed:
Ensure closing of open files

mvdbeek commented:
As a side note, there are many more data conversion tools that use unquoted paths, those should probably all be converted to single quoted paths.

bgruening commented on lib/galaxy/datatypes/converters/fastqcssangergz_to_fastqcssanger.xml:
```diff
@@ -0,0 +1,13 @@
+<tool id="CONVERTER_fastqcssangergz_to_fastqcssanger" name="Convert fastqcssanger.gz files to fastqcssanger" version="1.0.0" hidden="true">
+  <command>gzip -dcf '$input1' > '$output1'</command>
+  <inputs>
+    <page>
```
One more thing, why does this needs to have the `<page>` tag? It is deprecated isn't it?
ping @blankenberg

mvdbeek replied:
I had just taken this from the previous commits without further checking.
Conversion works fine without that tag.

blankenberg replied:
Yeah, don't need, and shouldn't use the page tag any more.

mvdbeek committed:
Remove deprecated page attribute

mvdbeek commented:
@galaxybot test this

mvdbeek commented:
Alright, anyone else any thoughts on this? Are we ready to merge?

abretaud commented:
+1 for me
I wonder if doing the same for bz2 files would be useful for anyone else than me?

mvdbeek commented:
> I wonder if doing the same for bz2 files would be useful for anyone else than me?

I'm sure it would be, at least having a converter from bzip2 to fastq. I think there are less tools that natively support bz2 though.

abretaud commented:
Ok, if it can wait a few days, I can make a PR on mvdbeek:fastqgz for this

abretaud committed:
add fastq(*).bz2 datatypes and converters

mvdbeek committed:
Merge pull request #1 from abretaud/bz2

Add fastq(*).bz2 datatypes and converters

mvdbeek commented:
awesome, thanks a lot @abretaud! (I love how a few days became 1)
@bgruening we still have your :+1: ?

abretaud commented:
>(I love how a few days became 1)

I needed to remove the post-it from my desk :)

abretaud committed:
Fix indentation and import (#2)

* fix indentation and import

* import fix

* refix indentation

* better looking imports

abretaud committed:
wrong import name (#3)

abretaud committed:
python lint

abretaud committed:
python lint

abretaud committed:
lint fix

bgruening commented:
:+1: from my site. It would be great if we can annotate the converters with requirement tags, but I think this can wait.

@jmchilton @dannon what do you think? Can we merge this?

mvdbeek commented:
> It would be great if we can annotate the converters with requirement tags, but I think this can wait.

Yep, I think we should do that, and actually test all the converter tools. But let's do that in another PR.

bgruening commented:
@jmchilton @dannon last call, before I merge this :)

jmchilton commented:
My silence isn't me ignoring this PR. I've looked at it several times. I really go back and forth on whether I'd like to see it merged or not. The devteam has known this is a real problem for years and years and has never done anything - it has never come up as a priority because of our particular file system on usegalaxy.org - so we should address this. I would just prefer to see it delt with at the framework level - but that would be large project. I think this cover the main use case for sequencing - I just worry about a flood of compressed types for things like proteomics where it is pretty common that tools deal with compressed files.

Please do merge @bgruening.

jmchilton commented:
Thanks @mvdbeek!

mvdbeek commented:
Thanks @jmchilton, and everyone involved in the PR!

peterjc commented:
Like @jmchilton I would also have preferred a generic compression flag allowing combination with any file format, but this is a welcome practical feature which ought to please many Galaxy Admins :)

mvdbeek commented:
> Like @jmchilton I would also have preferred a generic compression flag allowing combination with any file format, but this is a welcome practical feature which ought to please many Galaxy Admins :)

I think we're all on the same page in that regard ... I still hope we'll see this some day.
